src/Compilers/CSharp/Portable/Binder/Binder_NameConflicts.cs

```diff
@@ -65,16 +65,37 @@ internal void ValidateParameterNameConflicts(
                         continue;
                     }
 
-                    if (tpNames != null && tpNames.Contains(name))
+                    if (tpNames != null && tpNames.TryGetValue(name, out TypeParameterSymbol? tp))
                     {
-                        // CS0412: 'X': a parameter or local variable cannot have the same name as a method type parameter
-                        diagnostics.Add(ErrorCode.ERR_LocalSameNameAsTypeParam, GetLocation(p), name);
+                        if (tp.ContainingSymbol is NamedTypeSymbol { IsExtension: true })
+                        {
+                            if (p.ContainingSymbol != (object)tp.ContainingSymbol)
+                            {
+                                diagnostics.Add(ErrorCode.ERR_LocalSameNameAsExtensionTypeParameter, GetLocation(p), name);
+                            }
+                        }
+                        else if (p.ContainingSymbol is NamedTypeSymbol { IsExtension: true })
+                        {
+                            diagnostics.Add(ErrorCode.ERR_TypeParameterSameNameAsExtensionParameter, tp.GetFirstLocationOrNone(), name);
+                        }
+                        else
+                        {
+                            // CS0412: 'X': a parameter or local variable cannot have the same name as a method type parameter
+                            diagnostics.Add(ErrorCode.ERR_LocalSameNameAsTypeParam, GetLocation(p), name);
+                        }
                     }
 
                     if (!pNames.Add(name))
                     {
-                        // The parameter name '{0}' is a duplicate
-                        diagnostics.Add(ErrorCode.ERR_DuplicateParamName, GetLocation(p), name);
+                        if (parameters[0] is { ContainingSymbol: NamedTypeSymbol { IsExtension: true }, Name: var receiverName } && receiverName == name)
```

Comment:
>&& receiverName == name [](http://example.com/codeflow?start=129&length=23)

Consider adding a test to cover this part of the condition (`extension(int p1) void M(int p2, int p2)`)